File: backend/app/schemas/internship_workspace.py
"""Pydantic schemas for the Internship Workspace read / provisioning /
acceptance surface (database/migrations/038_internship_workspace.sql).

Phases 2-3 expose: a workspace summary + list, a detail view with the
PUBLISHED program preview, the provisioning-result body, and the
accept / decline / skill-selection request bodies.

Phase 5 adds the student assignment + submission surface: a list of the
workspace's published assignments with the student's latest attempt, an
assignment detail with full attempt history, and the create-submission
body. A submission is append-only (a resubmission is a new attempt).

Phase 6 adds the student's view of the industry's review of each of their
own attempts: verdict, feedback and score (NO reviewer identity -- the
student never sees who reviewed). Completion / certificate / stipend
schemas stay later phases.

`student_id` / `industry_id` are never accepted in a request -- every
endpoint derives identity from the authenticated token (require_student /
require_industry -> current_user.id) and RLS
(038_internship_workspace.sql) is the real access-control boundary.
"""

from typing import Literal

from pydantic import BaseModel, ConfigDict

# database/migrations/038_internship_workspace.sql -- CHECK value lists
WorkspaceStatus = Literal[
    "PENDING_ACCEPTANCE",
    "ACCEPTED",
    "IN_PROGRESS",
    "COMPLETED",
    "DECLINED",
    "RESCINDED",
]
WorkMode = Literal["REMOTE", "HYBRID"]
SkillRequirement = Literal["REQUIRED", "OPTIONAL"]
# database/migrations/037_internship_program.sql -- module_items.item_type
ModuleItemType = Literal["VIDEO", "PDF", "LINK", "TEXT"]
AssignmentType = Literal["ASSIGNMENT", "QUIZ", "PROJECT"]
SubmissionKind = Literal["LINK", "REPO", "FILE", "TEXT", "MIXED"]
SubmissionStatus = Literal[
    "SUBMITTED", "UNDER_REVIEW", "REVISION_REQUESTED", "ACCEPTED", "REJECTED"
]

# Mirrors app.services.internship_workspace_service.ProvisionResult.outcome
ProvisionOutcome = Literal[
    "CREATED",
    "ALREADY_EXISTS",
    "SKIPPED_WORK_MODE",
    "SKIPPED_NO_PROGRAM",
    "SKIPPED_NOT_SELECTED",
    "SKIPPED_NOT_INTERNSHIP",
]


class WorkspaceInternshipRef(BaseModel):
    """The internship posting behind a workspace. `status` is the
    posting's own lifecycle (DRAFT / PUBLISHED / CLOSED / ARCHIVED) --
    shown for context only, it is NOT an access gate.

    For the STUDENT endpoints this is resolved server-side (see
    internship_workspace_service) so the title/description stay available
    even after the posting is CLOSED / ARCHIVED, when student RLS on
    `internships` would otherwise hide the row. It exposes ONLY these
    fields -- never industry contact info or other posting columns."""

    id: str
    title: str | None = None
    description: str | None = None
    work_mode: str | None = None
    status: str | None = None


class InternshipWorkspaceSummary(BaseModel):
    id: str
    application_id: str
    internship_id: str
    student_id: str
    industry_id: str
    work_mode: str
    workspace_status: str
    accepted_at: str | None = None
    started_at: str | None = None
    completed_at: str | None = None
    declined_at: str | None = None
    decline_reason: str | None = None
    rescinded_at: str | None = None
    rescind_reason: str | None = None
    created_at: str | None = None
    updated_at: str | None = None
    internship: WorkspaceInternshipRef | None = None


class InternshipWorkspaceListResponse(BaseModel):
    workspaces: list[InternshipWorkspaceSummary]


# ---- program preview (read-only; students never author) ----


class WorkspaceProgramModuleItem(BaseModel):
    id: str
    title: str
    item_type: str
    content_url: str | None = None
    content_text: str | None = None
    order_index: int = 0


class WorkspaceProgramModule(BaseModel):
    id: str
    title: str
    description: str | None = None
    order_index: int = 0
    items: list[WorkspaceProgramModuleItem] = []


class WorkspaceProgramSkill(BaseModel):
    skill_id: str
    skill_name: str
    requirement: str  # REQUIRED | OPTIONAL


class WorkspaceProgramPreview(BaseModel):
    """The PUBLISHED internship_program for the workspace's internship,
    with its published modules / items / skills. Null when the industry
    has not published a program yet (the offer can still be accepted)."""

    id: str
    title: str
    summary: str | None = None
    status: str
    modules: list[WorkspaceProgramModule] = []
    skills: list[WorkspaceProgramSkill] = []


class InternshipWorkspaceDetail(InternshipWorkspaceSummary):
    program: WorkspaceProgramPreview | None = None
    # skill_ids the student has currently selected (all OPTIONAL program
    # skills; REQUIRED skills are always in scope and are not stored here).
    selected_skill_ids: list[str] = []


# ---- request bodies ----


class DeclineWorkspaceRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    reason: str | None = None


class SkillSelectionRequest(BaseModel):
    """The full desired set of OPTIONAL program skills (replace-set). An
    empty list clears the student's selections. Duplicates are normalised
    server-side."""

    model_config = ConfigDict(extra="forbid")

    skill_ids: list[str] = []


# ---- Phase 5: assignments + submissions / Phase 6: review outcome ----


class StudentSubmissionReview(BaseModel):
    """The industry's review of ONE of the student's own attempts, as the
    student sees it: the decision, the feedback and the score only. There
    is deliberately NO reviewer_id -- the student never sees who reviewed
    (submission_reviews' reviewer_id stays industry-side)."""

    verdict: str
    feedback: str | None = None
    score: float | None = None
    reviewed_at: str | None = None


class WorkspaceSubmissionResponse(BaseModel):
    """One of the student's OWN submission attempts, plus the industry's
    review of it (append-only history, newest first). The submission's own
    content is immutable once written -- a review never rewrites it."""

    id: str
    workspace_id: str
    assignment_id: str
    attempt_number: int
    submission_status: str
    repo_url: str | None = None
    live_url: str | None = None
    attachment_url: str | None = None
    notes: str | None = None
    submitted_at: str | None = None
    created_at: str | None = None
    updated_at: str | None = None
    reviews: list[StudentSubmissionReview] = []
    latest_review: StudentSubmissionReview | None = None


class WorkspaceAssignmentBase(BaseModel):
    id: str
    module_id: str
    program_id: str
    title: str
    description: str | None = None
    instructions: str | None = None
    assignment_type: str
    is_required: bool
    order_index: int
    due_offset_days: int | None = None
    submission_kind: str
    repo_required: bool
    live_url_expected: bool
    max_score: float | None = None
    linked_skill_id: str | None = None


class WorkspaceAssignmentSummary(WorkspaceAssignmentBase):
    module_title: str | None = None
    module_order_index: int = 0
    attempt_count: int = 0
    latest_submission: WorkspaceSubmissionResponse | None = None
    can_submit: bool = False
    submit_blocked_reason: str | None = None


class WorkspaceAssignmentListResponse(BaseModel):
    assignments: list[WorkspaceAssignmentSummary]


class WorkspaceAssignmentModuleRef(BaseModel):
    id: str
    title: str | None = None


class WorkspaceAssignmentDetail(BaseModel):
    assignment: WorkspaceAssignmentBase
    module: WorkspaceAssignmentModuleRef
    submissions: list[WorkspaceSubmissionResponse] = []
    attempt_count: int = 0
    can_submit: bool = False
    submit_blocked_reason: str | None = None


class CreateSubmissionRequest(BaseModel):
    """The student's work for one assignment. Only the fields
    workspace_submissions supports (migration 039); which are required
    depends on the assignment's submission_kind / repo_required /
    live_url_expected. `attempt_number` / `submission_status` are always
    server-set."""

    model_config = ConfigDict(extra="forbid")

    repo_url: str | None = None
    live_url: str | None = None
    attachment_url: str | None = None
    notes: str | None = None


# ---- provisioning ----


class ProvisionWorkspaceResponse(BaseModel):
    """Result of POST /api/v1/applications/{id}/provision-workspace.
    `workspace` is populated for CREATED / ALREADY_EXISTS, null for every
    no-op outcome."""

    outcome: ProvisionOutcome
    detail: str
    work_mode: str | None = None
    workspace: InternshipWorkspaceSummary | None = None

File: frontend/types/internship-workspace.ts
/**
 * Mirrors backend/app/schemas/internship_workspace.py -- field-for-field,
 * same nullability.
 *
 * The Internship Workspace is the student's post-selection internship
 * offer + training container (database/migrations/038_internship_workspace.sql).
 * There is one workspace per SELECTED internship application, for
 * REMOTE/HYBRID postings only. The student can accept or decline a
 * PENDING_ACCEPTANCE workspace and, once ACCEPTED, choose which OPTIONAL
 * program skills to focus on. Program authoring, assignments, submissions,
 * completion and the certificate are later phases.
 */

export type WorkspaceStatus =
  | "PENDING_ACCEPTANCE"
  | "ACCEPTED"
  | "IN_PROGRESS"
  | "COMPLETED"
  | "DECLINED"
  | "RESCINDED";

export const WORKSPACE_STATUSES: WorkspaceStatus[] = [
  "PENDING_ACCEPTANCE",
  "ACCEPTED",
  "IN_PROGRESS",
  "COMPLETED",
  "DECLINED",
  "RESCINDED",
];

/** Friendly label -- the backend value is never changed, only presented. */
export const WORKSPACE_STATUS_LABEL: Record<WorkspaceStatus, string> = {
  PENDING_ACCEPTANCE: "Pending Acceptance",
  ACCEPTED: "Accepted",
  IN_PROGRESS: "In Progress",
  COMPLETED: "Completed",
  DECLINED: "Declined",
  RESCINDED: "Rescinded",
};

export interface WorkspaceInternshipRef {
  id: string;
  title: string | null;
  description: string | null;
  work_mode: string | null;
  /** DRAFT / PUBLISHED / CLOSED / ARCHIVED -- context only, NOT an access
   * gate. The workspace stays readable regardless of this value. */
  status: string | null;
}

export interface InternshipWorkspaceSummary {
  id: string;
  application_id: string;
  internship_id: string;
  student_id: string;
  industry_id: string;
  work_mode: string;
  workspace_status: WorkspaceStatus;
  accepted_at: string | null;
  started_at: string | null;
  completed_at: string | null;
  declined_at: string | null;
  decline_reason: string | null;
  rescinded_at: string | null;
  rescind_reason: string | null;
  created_at: string | null;
  updated_at: string | null;
  internship: WorkspaceInternshipRef | null;
}

export type ModuleItemType = "VIDEO" | "PDF" | "LINK" | "TEXT";

export interface WorkspaceProgramModuleItem {
  id: string;
  title: string;
  item_type: string;
  content_url: string | null;
  content_text: string | null;
  order_index: number;
}

export interface WorkspaceProgramModule {
  id: string;
  title: string;
  description: string | null;
  order_index: number;
  items: WorkspaceProgramModuleItem[];
}

export interface WorkspaceProgramSkill {
  skill_id: string;
  skill_name: string;
  requirement: "REQUIRED" | "OPTIONAL";
}

export interface WorkspaceProgramPreview {
  id: string;
  title: string;
  summary: string | null;
  status: string;
  modules: WorkspaceProgramModule[];
  skills: WorkspaceProgramSkill[];
}

export interface InternshipWorkspaceDetail extends InternshipWorkspaceSummary {
  program: WorkspaceProgramPreview | null;
  /** OPTIONAL program skill_ids the student has currently selected.
   * REQUIRED skills are always in scope and are never listed here. */
  selected_skill_ids: string[];
}

// ---- Phase 5: assignments + the student's own submissions ----

export type SubmissionStatus =
  | "SUBMITTED"
  | "UNDER_REVIEW"
  | "REVISION_REQUESTED"
  | "ACCEPTED"
  | "REJECTED";

/** The industry's review of ONE of the student's own attempts, as the
 * student sees it. NO reviewer identity -- the student never learns who
 * reviewed. `reviewed_at` is the review row's timestamp. */
export interface StudentSubmissionReview {
  verdict: string;
  feedback: string | null;
  score: number | null;
  reviewed_at: string | null;
}

/** One of the student's OWN submission attempts, with the industry's
 * review of it (Phase 6 -- append-only history, newest first). The
 * submission's own content is immutable once written. */
export interface WorkspaceSubmission {
  id: string;
  workspace_id: string;
  assignment_id: string;
  attempt_number: number;
  submission_status: string;
  repo_url: string | null;
  live_url: string | null;
  attachment_url: string | null;
  notes: string | null;
  submitted_at: string | null;
  created_at: string | null;
  updated_at: string | null;
  reviews: StudentSubmissionReview[];
  latest_review: StudentSubmissionReview | null;
}

export interface WorkspaceAssignmentBase {
  id: string;
  module_id: string;
  program_id: string;
  title: string;
  description: string | null;
  instructions: string | null;
  assignment_type: string;
  is_required: boolean;
  order_index: number;
  due_offset_days: number | null;
  submission_kind: string;
  repo_required: boolean;
  live_url_expected: boolean;
  max_score: number | null;
  linked_skill_id: string | null;
}

export interface WorkspaceAssignmentSummary extends WorkspaceAssignmentBase {
  module_title: string | null;
  module_order_index: number;
  attempt_count: number;
  latest_submission: WorkspaceSubmission | null;
  can_submit: boolean;
  submit_blocked_reason: string | null;
}

export interface WorkspaceAssignmentList {
  assignments: WorkspaceAssignmentSummary[];
}

export interface WorkspaceAssignmentModuleRef {
  id: string;
  title: string | null;
}

export interface WorkspaceAssignmentDetail {
  assignment: WorkspaceAssignmentBase;
  module: WorkspaceAssignmentModuleRef;
  submissions: WorkspaceSubmission[];
  attempt_count: number;
  can_submit: boolean;
  submit_blocked_reason: string | null;
}

export interface CreateSubmissionInput {
  repo_url?: string | null;
  live_url?: string | null;
  attachment_url?: string | null;
  notes?: string | null;
}
